File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


typedef unsigned long long int64;

int64 cutLine(vector<int64> lines, int64 pivot) {

	int64 sum = 0;
	for (int64 line : lines)
		sum += line / pivot;

	return sum;
}

int main(void) {

	vector<int64> lines;

	int64 k;
	int64 n;

	cin >> k >> n;

	for (int64 i = 0; i < k; i++) {
		int64 line; cin >> line;
		lines.push_back(line);
	}

	sort(lines.begin(), lines.end());

	
	

	int64 start = 1;
	int64 end = lines[lines.size()-1];
	int64 mid;
	while (start < end) {

		if (end - start == 1) {
			int64 cutResult1 = cutLine(lines, start);
			int64 cutResult2 = cutLine(lines, end);

			if (cutResult2 < n)
				end = start;
			break;
		}

		int64 mid = (start + end) / 2;

		int64 cutResult = cutLine(lines,mid);

		if (cutResult >= n)
			start = mid;
		else
			end = mid - 1;
	}

	cout << end;
	
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

	static long lines[];
	
	public static long cutLine(long[] lines, long pivot) {
		long sum = 0;
		for (long line : lines)
			sum += line / pivot;

		return sum;
	}
	
	public static void main(String args[]) throws NumberFormatException, IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine()," ");
		
		int k = Integer.parseInt(st.nextToken());
		int n = Integer.parseInt(st.nextToken());
		
		lines = new long[k];
		
		for(int i = 0; i <k; i++)
			lines[i] = Integer.parseInt(br.readLine());
		
		Arrays.sort(lines);
		
		
		long start = 1;
		long end = lines[lines.length-1];
		long mid;
		while (start < end) {

			if (end - start == 1) {
				long cutResult1 = cutLine(lines, start);
				long cutResult2 = cutLine(lines, end);

				if (cutResult2 < n)
					end = start;
				break;
			}

			mid = (start + end) / 2;

			long cutResult = cutLine(lines,mid);

			if (cutResult >= n)
				start = mid;
			else
				end = mid - 1;
		}

		System.out.println(end);
		
	}
}